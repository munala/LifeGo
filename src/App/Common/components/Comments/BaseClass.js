/* eslint-disable react/no-unused-state */
import { Component } from 'react';

import propTypes from './propTypes';

class BaseClass extends Component {
  onChange = ({ target: { value } }) => {
    this.setState({
      comment: {
        ...this.state.comment,
        content: value,
      },
    });
  }

  onSubmit = async () => {
    const { comment } = this.state;
    this.setState(() => ({ submitting: true }));
    if (comment.content) {
      const { error } = await this.props.actions.addComment(this.props.bucketlist, comment);
      this.setState(() => ({ submitting: false }));
      if (!error) {
        this.setState(() => ({
          comment: {
            id: '',
            content: '',
          },
        }));
        this.setState(() => ({ typing: false }));
      }
    }
  }

  navigatePage = (direction) => {
    this.setState({
      page: this.state.page + (direction === 'next' ? 1 : -1),
    });
  }

  delete = () => {
    this.props.actions.deleteComment(
      this.props.bucketlist,
      this.state.selectedComment,
    );
  }

  focus = () => {
    this.setState({ typing: true });
  }

  editComment = () => {
    this.closeMenu();
    this.setState({
      comment: this.state.selectedComment,
      editMode: true,
      typing: true,
    });
  }

  cancel = () => {
    clearTimeout(this.timeout);
    this.setState({
      comment: {
        id: '',
        content: '',
      },
      editMode: false,
      snackOpen: false,
      deletedComment: { },
      selectedComment: {},
    });
  }

  saveComment = async () => {
    if (this.state.editMode) {
      const response = await this.props.actions.updateComment(
        this.props.bucketlist,
        this.state.comment,
      );
      if (!response.error) {
        this.setState({
          editMode: false,
          typing: false,
          comment: { content: '' },
          selectedComment: {},
        });
      }
    } else {
      this.onSubmit();
    }
  }

  deleteComment = (deletedComment) => {
    this.setState(
      { deletedComment },
      () => {
        this.setState({ snackOpen: true });
        this.timeout = setTimeout(() => {
          this.delete(deletedComment);
          this.setState(
            { snackOpen: false },
            () => {
              setTimeout(() => {
                this.setState({
                  deletedComment: { },
                  selectedComment: { name: '', content: '' },
                });
              }, 1000);
            },
          );
        }, 5000);
      },
    );
  }

  editComment = (comment) => {
    this.handleClose();
    this.setState({
      comment,
      editMode: true,
    });
  }

  handleClick = ({ event, selectedComment }) => {
    this.setState({
      selectedComment,
      anchorEl: event.currentTarget,
    });
  };

  selectComment = ({ menuItem }) => {
    const { selectedComment: comment } = this.state;
    const actions = {
      Edit: () => this.editComment(comment),
      Delete: () => this.deleteComment(comment),
    };
    actions[menuItem]();
    this.handleClose();
  }

  handleClose = () => {
    this.setState({ anchorEl: null });
  };
}

BaseClass.propTypes = propTypes;

export default BaseClass;
