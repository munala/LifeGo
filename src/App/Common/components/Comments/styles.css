.comments-container {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: stretch;
  font-family: sans-serif;
  font-size: 12px;
  flex-grow: 1;
}

.comments {
  width: 100%;
  overflow: scroll;
  flex-direction: column;
  justify-content: flex-start;
  align-items: flex-end;
}

.comment-container {
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: flex-start;
  margin: 5px;
}


.comment-body {
  background-color: #f7f7f7;
  border-radius: 10px;
  overflow: hidden;
  padding: 10px;
  margin-left: 5px;
}

.comment-user {
  color: #555;
  font-weight: bold;
  white-space: nowrap;
  cursor: pointer;
}

.comment-content {
  font-size: 12px;
  color: #666;
  margin-left: 5px;
}

.time {
  display: flex;
  color: #ddd;
  font-size: 10px;
  align-self: flex-start;
  margin-left: 10px;
  margin-bottom: 5px;
}

.nav-link {
  margin: 5px;
}
.nav-link:hover {
  text-decoration: underline;
}

a {
  text-decoration: none;
  color: #777
}

.empty-text {
  display: flex;
  justify-content: center;
  color:#666;
  text-align: center;
}
