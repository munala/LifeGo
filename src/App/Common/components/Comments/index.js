import React from 'react';
import PropTypes from 'prop-types';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import MoreHorizIcon from '@material-ui/icons/MoreHoriz';
import IconButton from '@material-ui/core/IconButton';

import SnackBarComponent from '../SnackBarComponent';
import Avatar from '../Avatar';
import NewInput from '../NewInput';
import BaseClass from './BaseClass';
import avatar from '../../../../assets/images/user.png';
import { setTime } from '../../../../utils';
import './styles.css';

const ITEM_HEIGHT = 48;

class Comments extends BaseClass {
  state = {
    page: 0,
    comment: {
      content: '',
    },
    selectedComment: {
      id: '',
      content: '',
    },
    deletedComment: {},
    typing: false,
    submitting: false,
    editMode: false,
    snackOpen: false,
    hoveredComment: {},
  }

  renderComments = ({ comments, mode, goToProfile }) => comments
    .filter(comment => comment.id !== this.state.deletedComment.id)
    .slice(this.state.page * 8, (this.state.page * 8) + 8)
    .map((comment) => {
      const { bucketlist, profile } = this.props;
      const menuItems = comment.senderId === profile.id ? ['Edit', 'Delete'] : ['Delete'];
      const {
        anchorEl, selectedComment, snackOpen, hoveredComment,
      } = this.state;

      return (
        <div
          key={comment.id}
          className="comment-container"
          onMouseEnter={() => this.hover(comment)}
          onMouseLeave={() => this.hover(comment)}
        >
          <Avatar
            src={comment.userPictureUrl
              ? comment.userPictureUrl
              : avatar // eslint-disable-line global-require
            }
            style={{ display: 'flex', height: 30, width: 30 }}
          />
          <div className="comment">
            <div className="comment-left">
              <div className="comment-body">
                <span onClick={() => goToProfile({ id: comment.senderId })} className="comment-user">
                  {comment.user}
                </span>
                <span className="comment-content">
                  {comment.content}
                </span>
              </div>
              {
                (bucketlist.userId === profile.id || comment.senderId === profile.id)
                && !snackOpen && hoveredComment.id === comment.id ?
                  <IconButton
                    aria-label="More"
                    aria-owns={anchorEl ? 'long-menu' : null}
                    aria-haspopup="true"
                    style={{ maxHeight: 20, alignSelf: 'center' }}
                    onClick={event => this.handleClick({ event, selectedComment: comment })}
                  >
                    <MoreHorizIcon style={{ color: '#444' }} />
                  </IconButton> :
                  <div className="icon-placeholder" />
              }
              {
                selectedComment.id === comment.id &&
                <Menu
                  id="long-menu"
                  anchorEl={anchorEl}
                  open={Boolean(anchorEl)}
                  onClose={this.handleClose}
                  PaperProps={{
                    style: {
                      maxHeight: ITEM_HEIGHT * 4.5,
                      paddingLeft: 10,
                      paddingRight: 10,
                    },
                  }}
                >
                  {menuItems.map(menuItem => (
                    <MenuItem key={menuItem} onClick={() => this.selectComment({ menuItem })}>
                      {menuItem}
                    </MenuItem>
                  ))}
                </Menu>
              }
            </div>
            <div className="time">
              {mode && `${setTime(comment).createdAt}${setTime(comment).time}`}
            </div>
          </div>
        </div>
      );
    });

  render() {
    const {
      bucketlist: {
        comments,
      }, goToProfile, mode, profile,
    } = this.props;

    const {
      page, comment, submitting, editMode, snackOpen, deletedComment,
    } = this.state;

    const { length } = comments.filter(comm => comm.id !== deletedComment.id);

    const lastPage = Math.floor(comments.length / 8);

    return (
      <div className="comments-container">
        <div className="comments">
          {
            length > 0 && page > 0 &&
            <div className="nav-link" onClick={() => this.navigatePage('previous')}>
              <a href="#">more comments</a>
            </div>
          }

          {this.renderComments({ comments, goToProfile, mode })}

          {
            length > 0 && page < lastPage &&
            <div className="nav-link" onClick={() => this.navigatePage('next')}>
              <a href="#">previous comments</a>
            </div>
          }
        </div>
        <NewInput
          name="comment"
          pictureUrl={profile.pictureUrl}
          content={{ ...comment, type: 'comment' }}
          onChange={this.onChange}
          save={this.saveComment}
          submitting={submitting}
          buttonLabel={editMode ? 'save' : 'post'}
        />
        <SnackBarComponent
          open={snackOpen}
          message={{
            content: `You have deleted ${deletedComment && deletedComment.user}'s comment`,
            success: true,
          }}
          closeSnackBar={this.cancel}
          undo="undo"
        />
      </div>
    );
  }
}

Comments.propTypes = {
  bucketlist: PropTypes.shape({
    name: PropTypes.string,
    id: PropTypes.number,
    createdAt: PropTypes.string,
    updatedAt: PropTypes.string,
    description: PropTypes.string,
    comments: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string,
      id: PropTypes.number,
      createdAt: PropTypes.string,
      updatedAt: PropTypes.string,
      done: PropTypes.bool,
    })),
    items: PropTypes.arrayOf(PropTypes.shape({
      content: PropTypes.string,
      id: PropTypes.number,
      createdAt: PropTypes.string,
      updatedAt: PropTypes.string,
    })),
  }).isRequired,
  actions: PropTypes.shape({
    addComment:
    PropTypes.func.isRequired,
    deleteComment: PropTypes.func.isRequired,
  }).isRequired,
  profile: PropTypes.shape({
    id: PropTypes.number,
    username: PropTypes.string,
    displayName: PropTypes.string,
    email: PropTypes.string,
    pictureUrl: PropTypes.string,
    friends: PropTypes.arrayOf(PropTypes.shape({})),
    searchUsers: PropTypes.arrayOf(PropTypes.shape({})),
  }).isRequired,
  mode: PropTypes.string,
  goToProfile: PropTypes.func.isRequired,
};

export default Comments;
