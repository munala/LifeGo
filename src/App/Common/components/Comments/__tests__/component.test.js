import React from 'react';
import { shallow } from 'enzyme';
import Comments from '../';

describe('Comments tests', () => {
  let wrapper;
  const addComment = jest.fn().mockReturnValue({});
  const deleteComment = jest.fn().mockReturnValue({});
  const goToProfile = jest.fn();
  const bucketlist = { id: 1, comments: [{ id: 1, content: 'comment' }] };
  const openDialog = jest.fn();

  beforeAll(() => {
    wrapper = shallow(<Comments
      bucketlist={bucketlist}
      actions={{
        addComment,
        deleteComment,
      }}
      profile={{ id: 1 }}
      mode="single"
      goToProfile={goToProfile}
      openDialog={openDialog}
    />);
  });

  it('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('handles text change', () => {
    const input = wrapper.find('NewInput');
    input.props().onChange({ target: { value: 'value' } });
    expect(wrapper.instance().state.comment.content).toEqual('value');
  });

  it('handles submitting', () => {
    const input = wrapper.find('NewInput');
    input.props().onChange({ target: { value: 'value' } });

    input.props().save();

    expect(addComment).toHaveBeenCalledTimes(1);
    expect(addComment).toHaveBeenCalledWith(bucketlist, { content: 'value' });
  });

  it('handles going to profile', () => {
    const comment = wrapper.find('a');

    comment.props().onClick();

    expect(goToProfile).toHaveBeenCalledTimes(1);
  });
});
