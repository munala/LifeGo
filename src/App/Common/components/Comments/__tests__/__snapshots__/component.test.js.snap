// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments tests renders without crashing 1`] = `
<div
  className="comments-container"
>
  <div
    className="comments"
  >
    <div
      className="comment-container"
      key="1"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <UserAvatar
        name="user avatar"
        onClick={[Function]}
        src="user.png"
        style={
          Object {
            "display": "flex",
            "height": 30,
            "width": 30,
          }
        }
      />
      <div
        className="comment"
      >
        <div
          className="comment-left"
        >
          <div
            className="comment-body"
          >
            <span
              className="comment-user"
              onClick={[Function]}
            />
            <span
              className="comment-content"
            >
              comment
            </span>
          </div>
          <div
            className="icon-placeholder"
          />
        </div>
        <div
          className="time"
        >
          August 17th
        </div>
      </div>
    </div>
  </div>
  <NewInput
    buttonLabel="post"
    cancel={[Function]}
    content={
      Object {
        "content": "",
        "type": "comment",
      }
    }
    editMode={false}
    focus={[Function]}
    name="comment"
    onChange={[Function]}
    pictureUrl={null}
    save={[Function]}
    submitting={false}
  />
  <SnackbarComponent
    closeSnackBar={[Function]}
    message={
      Object {
        "content": "You have deleted undefined's comment",
        "success": true,
      }
    }
    open={false}
    undo="undo"
  />
</div>
`;
